csrc/scheduler/normalization_inner_outer_tma_ws.cpp

```diff
@@ -193,85 +137,51 @@ void getHeuristics(
-        get_heuristics_given_vect_threads(vect_factor, threads_per_block));
-  }
+  int64_t bdimx = std::min(128L, after_vect);
+  bdimx = std::max(bdimx, ceilDiv(after_vect, max_persistent_batch));
+  bdimx = scheduler_utils::roundUpToN(bdimx, 128L);
 
-  // Sort the heuristics based on the register usage and occupancy.
-  std::stable_sort(
-      iop_candidates.begin(),
-      iop_candidates.end(),
-      [](const InnerOuterParams& a, const InnerOuterParams& b) {
-        // If a thread can use more registers than required, there is a high
-        // chance that it can avoid register spilling and compiler can optimize
-        // for better instruction level parallelism.
-        int64_t extra_regs_a =
-            a.available_register_per_thread - a.required_register_per_thread;
-        int64_t extra_regs_b =
-            b.available_register_per_thread - b.required_register_per_thread;
-        if (extra_regs_a > 0 && extra_regs_b < 0) {
-          return true;
-        } else if (extra_regs_a < 0 && extra_regs_b > 0) {
-          return false;
-        }
-        // High occupancy provides better threads level parallelism.
-        // 25% is sufficient since ILP is high due to persistent batch sizes
-        // which is equivalent to unrolling inner dim.
-        if (a.warps_per_sm != b.warps_per_sm &&
-            (a.warps_per_sm < 16 || b.warps_per_sm < 16)) {
-          return a.warps_per_sm > b.warps_per_sm;
-        }
-        // Tie breaker, smaller threads_per_block to reduce communication
-        // overhead
-        return a.threads_per_block < b.threads_per_block;
-      });
-
-  // Pick the best heuristic
-  auto iop = iop_candidates.front();
-  rparams->block_dim_inner_reduction_extra = ParallelType::TIDy;
+  auto iop = get_heuristics_given_vect_threads(vect_factor, bdimx);
   rparams->combined_split_grid_inner_dim =
       iop.vectorization_factor_outer * iop.bdimx * iop.gdimy < inner_dim_numel;
-  rparams->static_bdimx = true;
-  rparams->static_bdimy = true;
-  iop.bdimz = ceilDiv(
-      ceilDiv(ceilDiv(inner_dim_numel / iop.inner_vect, iop.bdimx), iop.bdimy),
-      iop.inner_batch);
-  NVF_ERROR(iop.bdimz == 1, "bdimz must be 1.");
 
   // check all the parameters in InnerOuterParams are set.
   iop.verify();
 
   rparams->persistent_kernel = true;
   rparams->fastest_dim = true;
   rparams->combined_inner_outer = true;
-  // tmp_gmem is the intermediate result of outer reduction, its dtype is float,
-  // so the maximum vectorization factor is 4.
+
+  // Set circular buffer, n_stages and n_prefetch are tunable parameters.
+  // n_stages is also limited by smem
+  int64_t max_n_copies = (int64_t)dev_prop->sharedMemPerMultiprocessor /
+      (smem_overhead + smem_buffer_size);
+  int64_t iter_remaining = ceilDiv(outer_dim_numel, iop.gdimy);
+  int64_t n_stages_prefered = std::min(2L, iter_remaining);
+  int64_t n_stages_max_allowed = max_n_copies;
+  int64_t n_stages = std::min(n_stages_prefered, n_stages_max_allowed);
```

Comment:
```suggestion
  int64_t n_stages = std::min(n_stages_prefered, max_n_copies);
```
Would prefer a code comment rather then a renamed variable for added clarity.